src/libcore/task/wake.rs

```diff
@@ -63,21 +65,131 @@ pub struct RawWakerVTable {
+    /// resources that are associated with this instance of a [`RawWaker`] and
+    /// associated task.
+    wake: unsafe fn(*const ()),
+
+    /// This function will be called when `wake_by_ref` is called on the [`Waker`].
+    /// It must wake up the task associated with this [`RawWaker`].
+    ///
+    /// This function is similar to `wake`, but must not consume the provided data
     /// pointer.
-    pub wake: unsafe fn(*const ()),
+    wake_by_ref: unsafe fn(*const ()),
 
     /// This function gets called when a [`RawWaker`] gets dropped.
     ///
     /// The implementation of this function must make sure to release any
     /// resources that are associated with this instance of a [`RawWaker`] and
     /// associated task.
-    pub drop: unsafe fn(*const ()),
+    drop: unsafe fn(*const ()),
+}
+
+impl RawWakerVTable {
+    /// Creates a new `RawWakerVTable` from the provided `clone`, `wake`, and
+    /// `drop` functions.
+    ///
+    /// # `clone`
+    ///
+    /// This function will be called when the [`RawWaker`] gets cloned, e.g. when
+    /// the [`Waker`] in which the [`RawWaker`] is stored gets cloned.
+    ///
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a [`RawWaker`] and associated
+    /// task. Calling `wake` on the resulting [`RawWaker`] should result in a wakeup
+    /// of the same task that would have been awoken by the original [`RawWaker`].
+    ///
+    /// # `wake`
+    ///
+    /// This function will be called when `wake` is called on the [`Waker`].
+    /// It must wake up the task associated with this [`RawWaker`].
+    ///
+    /// The implementation of this function must make sure to release any
+    /// resources that are associated with this instance of a [`RawWaker`] and
+    /// associated task.
+    ///
+    /// # `wake_by_ref`
+    ///
+    /// This function will be called when `wake_by_ref` is called on the [`Waker`].
+    /// It must wake up the task associated with this [`RawWaker`].
+    ///
+    /// This function is similar to `wake`, but must not consume the provided data
+    /// pointer.
+    ///
+    /// # `drop`
+    ///
+    /// This function gets called when a [`RawWaker`] gets dropped.
+    ///
+    /// The implementation of this function must make sure to release any
+    /// resources that are associated with this instance of a [`RawWaker`] and
+    /// associated task.
+    #[rustc_promotable]
+    #[cfg_attr(stage0, unstable(feature = "futures_api_const_fn_ptr", issue = "50547"))]
+    #[cfg_attr(not(stage0), stable(feature = "futures_api", since = "1.35.0"))]
+    // `rustc_allow_const_fn_ptr` is a hack that should not be used anywhere else
+    // without first consulting with T-Lang.
+    //
+    // FIXME: remove whenever we have a stable way to accept fn pointers from
+    // const fn (see rfcs/pull/2632)
+    #[cfg_attr(not(stage0), rustc_allow_const_fn_ptr)]
```

Comment:
I am rescinding my support for this hack (see also https://github.com/rust-lang/rfcs/pull/2632#issuecomment-483390877) until the `?const` discussion has been resolved. The only reason @Centril and I signed off on it was because we were under the impression that there will be some sort of opt-out of the constness of function pointers and trait bounds on `const fn`s.

If `rustc_allow_const_fn_ptr` stays as a "permanent hack" I will be very unhappy (mostly because I see no reason not to have something like it usable by users on stable).

Reply:
Clearly we *want* to have something which adds support for this feature to the language, yes? If so, I don't understand why we would hold off on allowing this.

Reply:
In particular, the stabilization of `futures_api` *depends* on having this attribute-- I'd strongly consider taking the importance of that feature and its timeline into account here.

Reply:
> Clearly we _want_ to have something which adds support for this feature to the language, yes?

I thought so too and it was based on this understanding that me and Oliver agreed to the attribute. However, @withoutboats's left a concern about it in https://github.com/rust-lang/rfcs/pull/2632#issuecomment-483299150 making it highly unclear that we "clearly want to have something...".

Reply:
To be clear, I am completely comfortable committing to having some mechanism which will allow fn pointers to be passed as arguments to const fns, which is all that this function commits us to.

Reply:
I also want to say find this comment thread extremely stress inducing.

Reply:
In [this comment](https://github.com/rust-lang/rfcs/pull/2632#issuecomment-483421520), @withoutboats outlined two alternatives that they support, either of which would still allow this function (which doesn't call the `fn` pointers it is passed).

Reply:
I don't even *prefer* the alternatives I enumerated; I just want more experience and user feedback with const fn before committing to new syntax like `?const`. There are many shapes a solution which supports writing RawWaker::new on stable could take, and I don't think we know what the best one is yet. That's all!

Reply:
> To be clear, I am completely comfortable committing to having some mechanism which will allow fn pointers to be passed as arguments to const fns, which is all that this function commits us to.

> I don't even _prefer_ the alternatives I enumerated; I just want more experience and user feedback with const fn before committing to new syntax like `?const`. There are many shapes a solution which supports writing RawWaker::new on stable could take, and I don't think we know what the best one is yet. That's all!

This clarification is very helpful and is all I really wanted to be certain of. 

I agree with collecting more user data wrt. syntax, but it seems to me that if you want to do that, then experimentation on nightly is preferable.

Reply:
Cool-- it sounds like we're all still okay with `#[rustc_allow_const_fn_ptr]` then? I'll resolve this thread in that case, and we can continue discussion about `?const` and nightly experimentation in the related RFC thread.

Reply:
I apologize for my comment. I was very agitated by the comment on the trait-bounds-on-const-fn that to me seemed to question the existance of the entire class of features that makes the RFC useful in my book. I'm very sorry about the unnecessary escalation.

Reply:
@oli-obk Oops, I accidentally put two things into one comment thread, which @cramertj marked as resolved, otherwise you might've seen it.

I think this is fine *and unrelated to* #2632 because it can be expressed with pattern aliases: https://github.com/rust-lang/rust/pull/59739#discussion_r272793688